### PR TITLE
[🦩] Add ASTGen.

### PR DESCRIPTION
**Pull request by zoecarver**
Refs https://github.com/apple/swift-syntax/pull/703

### CODE REVIEW

**Comment by zoecarver** [comment]
Thanks @CodaFi for the initial code that this was built off of.

File: lib/Frontend/CompilerInvocation.cpp [hunk 1]
```diff
@@ -842,6 +842,7 @@ static bool ParseLangArgs(LangOptions &Opts, ArgList &Args,
   }
 
   Opts.EnableCXXInterop |= Args.hasArg(OPT_enable_experimental_cxx_interop);
+  Opts.EnableSwiftParser |= Args.hasArg(OPT_enable_experimental_swift_parser);
```
Comment (DougGregor) [inline lib/Frontend/CompilerInvocation.cpp:845]: Won't be needed if you introduce this as an experimental feature

File: lib/Syntax/CMakeLists.txt [hunk 2]
```diff
@@ -4,19 +4,19 @@ else()
   set(SWIFT_GYB_FLAGS --line-directive "\'#line" "%(line)d" "\"%(file)s\"\'")
 endif()
 
-add_swift_host_library(swiftSyntax STATIC
+add_swift_host_library(legacySwiftSyntax STATIC
```
Comment (DougGregor) [inline lib/Syntax/CMakeLists.txt:7]: Let's not end up like the legacy pass manager, eh? ;)

Comment (DougGregor) [reply lib/Syntax/CMakeLists.txt:7]: I don't know if we actually need this, given that CMake has scoping for targets. But it's fine if we need to do it---this should be going away soonish.

File: tools/driver/CMakeLists.txt [hunk 3]
```diff
@@ -49,6 +49,9 @@ add_swift_host_tool(swift-frontend
 )
 target_link_libraries(swift-frontend
                       PUBLIC
+                        ASTGen
+                        SwiftParser
+                        SwiftSyntax
```
Comment (DougGregor) [inline tools/driver/CMakeLists.txt:54]: Support for linking in SwiftSyntax and SwiftParser was already introduced in https://github.com/apple/swift/pull/60888. Can we ride on top of that?

File: lib/Parse/CMakeLists.txt [hunk 4]
```diff
@@ -6,6 +6,9 @@ else()
   set(SWIFT_GYB_FLAGS --line-directive "\'#line" "%(line)d" "\"%(file)s\"\'")
 endif()
 
+add_subdirectory(${SWIFT_SOURCE_DIR}/../swift-syntax ${CMAKE_CURRENT_BINARY_DIR}/swift-syntax)
```
Comment (DougGregor) [inline lib/Parse/CMakeLists.txt:9]: We already have `SWIFT_PATH_TO_EARLYSWIFTSYNTAX_BUILD_DIR` for this, which uses exported targets. Please build on top of that infrastructure.

File: lib/AST/CASTBridging.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+
+using namespace swift;
+
+template <typename T>
+inline llvm::ArrayRef<T> getArrayRef(BridgedArrayRef bridged) {
+  return {static_cast<const T *>(bridged.data), size_t(bridged.numElements)};
+}
+
+BridgedIdentifier
+SwiftASTContext_getIdentifier(void *ctx, const char *_Nullable str, long len) {
```
Comment (DougGregor) [inline lib/AST/CASTBridging.cpp:20]: This whole file is mind-numbingly dull to write, but at least it's straightforward!

Comment (zoecarver) [reply lib/AST/CASTBridging.cpp:20]: Yeah. I wonder if we could write a python script that gets us 90% of the way there.

File: tools/driver/CMakeLists.txt [hunk 6]
```diff
@@ -49,9 +49,15 @@ add_swift_host_tool(swift-frontend
 )
 target_link_libraries(swift-frontend
                       PUBLIC
+                        ASTGen
+                        swiftCore # TODO: how do we fix this?
```
Comment (zoecarver) [inline tools/driver/CMakeLists.txt:53]: @DougGregor any idea how to get the link paths for this?

**Comment by zoecarver** [comment]
I updated this to use your changes from https://github.com/apple/swift/pull/60888. Much cleaner :)

**Comment by zoecarver** [comment]
@swift-ci please test.

**Comment by zoecarver** [comment]
@swift-ci please test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

File: include/swift/AST/CASTBridging.h [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+void *SimpleIdentTypeRepr_create(void *ctx, void *loc, const char *id);
+
+void *UnresolvedDotExpr_create(void *ctx, void *base, void *dotLoc, const char *name, void *nameLoc);
+
+void *ClosureExpr_create(void *ctx, void *body, void *dc);
+
+void NominalTypeDecl_setMembers(void *decl, BridgedArrayRef members);
+
+struct DeclContextAndDecl {
```
Comment (zoecarver) [inline include/swift/AST/CASTBridging.h:133]: This type really sucks :(

**Comment by zoecarver** [comment]
@swift-ci please smoke test

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test

**Comment by zoecarver** [comment]
@swift-ci please smoke test

**Comment by zoecarver** [comment]
@DougGregor pretty sure that resolved all the linker errors. I want to see the CI pass once, then I'll add the cmake option.

**Comment by zoecarver** [comment]
Closing in favor of https://github.com/apple/swift/pull/61426#pullrequestreview-1129153523

**Pull request closed without merging**